core/src/banking_stage/consumer.rs

```diff
@@ -479,6 +479,32 @@ impl Consumer {
             })
         ));
 
+        // Remove reserved block space for transaction failed to acquire lock
+        let unlocked_cost_results: Vec<_> = batch
+            .lock_results()
+            .iter()
+            .zip(transaction_qos_cost_results.iter_mut())
+            .map(|(lock_result, cost)| {
+                if let Err(lock_err) = lock_result {
+                    if cost.is_ok() {
+                        // set cost to lock_err, so it won't be  accidentally removed more than once
+                        let mut c = Err(lock_err.clone());
+                        std::mem::swap(cost, &mut c);
+                        return Some(c.unwrap());
+                    }
+                }
+                None
+            })
+            .collect();
```

Comment:
we don't need to allocate a new vec here, we can just iterate over our qos and lock result pairs...something like
```rust
        // Remove reserved block space for transaction failed to acquire lock
        let mut cost_tracker = bank.write_cost_tracker().unwrap();
        transaction_qos_cost_results
            .iter_mut()
            .zip(batch.lock_results())
            .filter_map(|(qos_cost_result, lock_result)| {
                match (qos_cost_result.as_ref(), lock_result) {
                    (Ok(_), Err(lock_err)) => Some((qos_cost_result, lock_err)),
                    _ => None,
                }
            })
            .for_each(|(qos_cost_result, lock_err)| {
                cost_tracker.remove(qos_cost_result.as_ref().unwrap());
                *qos_cost_result = Err(lock_err.clone());
            });
```

Reply:
TIL! I was having trouble to change `qos_cost_result` while it is borrowed by `match` (i think).

Before making the change tho, are we good with this approach, or would like to "lock-reserve-unlock" you mentioned elsewhere?